ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1841,6 +1908,60 @@ void addChunkId(StoreKey storeKey, long chunkSize, int chunkIndex) {
       indexToChunkIdsAndChunkSizes.put(chunkIndex, new Pair<>(storeKey, chunkSize));
     }
 
+    /**
+     * Validate the reserved metadata chunk id is passed in correctly where it needs to be specified.
+     * @throws RouterException in case of any errors.
+     */
+    private void validateReservedMetadataChunkId() throws RouterException {
+      if (!options.isChunkUpload() && Objects.nonNull(passedInBlobProperties.getReservedMetadataBlobId())) {
+        reservedMetadataIdMetrics.reservedMetadataPassedInForNonChunkedUploadCount.inc();
+        throw new RouterException("Reserved metadata should be passed in for request that isn't chunked upload.",
+            RouterErrorCode.InvalidOrMismatchedStitchBlobReservedMetadataChunkId);
+      }
+      if (options.isChunkUpload() && passedInBlobProperties.getReservedMetadataBlobId() == null) {
+        reservedMetadataIdMetrics.noReservedMetadataForChunkedUploadCount.inc();
+        throw new RouterException("No reserved metadata passed in for chunked upload",
```

Comment:
Don't we have a config to enable the reservation? If it is set to false, should expect reserved metadata id?

Reply:
`validateReservedMetadataChunkId` will always throw an exception for invalid cases. The caller can then make a determination how it wants to handle the exception.
In this case, the `MetadataPutChunk` constructor that calls it, handles the exception based on whether config is enabled or not.